packages/types/src/transaction/get-stub-action-view-from-plan.ts

```diff
@@ -0,0 +1,122 @@
+import {
+  ActionPlan,
+  ActionView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/transaction/v1alpha1/transaction_pb';
+import {
+  Value,
+  ValueView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/asset/v1alpha1/asset_pb';
+import {
+  Note,
+  NoteView,
+  OutputPlan,
+  OutputView,
+  SpendPlan,
+  SpendView,
+} from '@buf/penumbra-zone_penumbra.bufbuild_es/penumbra/core/component/shielded_pool/v1alpha1/shielded_pool_pb';
+import { JsonValue } from '@bufbuild/protobuf';
+import { uint8ArrayToBase64 } from '../base64';
+
+const getValueView = (
+  value: Value | undefined,
+  denomMetadataByAssetId: Record<string, JsonValue>,
+): ValueView => {
+  if (!value) throw new Error('No value to view');
+  if (!value.assetId) throw new Error('No asset ID in value');
+  if (!value.amount) throw new Error('No amount in value');
+
+  const denomMetadata = denomMetadataByAssetId[uint8ArrayToBase64(value.assetId.inner)];
+  if (!denomMetadata) throw new Error('Asset ID in spend plan refers to an unknown asset type');
+
+  return new ValueView({
+    valueView: {
+      case: 'knownDenom',
+      value: {
+        amount: value.amount,
+        denom: denomMetadata,
+      },
+    },
+  });
+};
+
+const getNoteView = (note: Note | undefined, denomMetadataByAssetId: Record<string, JsonValue>) => {
+  if (!note) throw new Error('No note to view');
+  if (!note.address) throw new Error('No address in note');
+  if (!note.value) throw new Error('No value in note');
+
+  return new NoteView({
+    address: {
+      addressView: {
+        case: 'opaque',
+        value: { address: note.address },
+      },
+    },
+    value: getValueView(note.value, denomMetadataByAssetId),
+  });
+};
+
+const getSpendView = (
+  spendPlan: SpendPlan,
+  denomMetadataByAssetId: Record<string, JsonValue>,
+): SpendView => {
+  if (!spendPlan.note?.address) throw new Error('No address in spend plan');
+
+  return new SpendView({
+    spendView: {
+      case: 'visible',
+      value: {
+        note: getNoteView(spendPlan.note, denomMetadataByAssetId),
+      },
+    },
+  });
+};
+
+const getOutputView = (
+  outputPlan: OutputPlan,
+  denomMetadataByAssetId: Record<string, JsonValue>,
+): OutputView => {
+  if (!outputPlan.destAddress) throw new Error('No destAddress in output plan');
+
+  return new OutputView({
+    outputView: {
+      case: 'visible',
+
+      value: {
+        note: {
+          value: getValueView(outputPlan.value, denomMetadataByAssetId),
+          address: {
+            addressView: {
+              case: 'opaque',
+              value: {
+                address: outputPlan.destAddress,
```

Comment:
@hdevalence is this the correct mapping of the output plan's `destAddress` to the `outputView`? The `outputPlan` has a _top-level_ `destAddress` property next to a top-level `value` property, and I'm putting them both into the `note` of this `OutputView` instance. That feels wrong to me somehow, but I don't know enough about these object structures yet to know if that's incorrect. So I want to make sure I'm not misrepresenting the action to the user.

(This PR in general isn't ready for review yet, but I wanted to make this comment before I forget.)

Reply:
also CC @turbocrime

Reply:
Yes, this is actually correct. The `OutputPlan` contains all the data that will be used to generate the note.

The `Output` action consists of:

- a note commitment to the newly created note
- an encryption of the note itself, decryptable by both the sender and receiver
- a zk proof that the note commitment was correctly formed

So the reason that the output plan has the data separately (value, address) is that that data is used for each of those parts. 

Here where we're making only a "skeletonized" version, this is the correct translation.